### PR TITLE
JAMES-3365 JMAP back reference resolution

### PR DESCRIPTION
**Pull request by chibenwa**
**Strategy**: before parsing ones method argument, rely on **play-json** JSON transformers to substitute back-references in the supplied arguments by the results of previously executed methods.

Advantage: 
 - this method is **fully** transparent to method developer, who just don't need to worry about back-reference resolving, else than aat argument parsing time, and recording result time. => Easy!
 - this method allows complex and arbitrary client supplied value. We don't need to guess one use cases

Disadantage:
 - No strong typing: one can for instance pass emailIds into mailboxIds

**Notes**: 
 - This work aims at demonstrating this approach works, and did not at all address error management (which might require heroic monad management skills).
 - I did not reuse existing JSON pointer lib given that JMAP syntax slightly differs from the spec.
 - I'm unsure the /xx/*/yy/*zz work as is...

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala [hunk 1]
```diff
@@ -80,46 +86,67 @@ class ProcessingContext {
...
- private def resolveBackReferenceNeverFails(serializer: Serializer)(entry: (String, JsValue)): (String, JsValue) = {
-  val newEntry: (String, JsResult[JsValue]) = resolveBackReference(serializer)(entry)
-  // if (newEntry._2.isError) {
-  (newEntry._1, newEntry._2.get)
- }
+  private def resolveBackReferences(serializer: Serializer, array: collection.IndexedSeq[JsValue]): JsResult[JsValue] = {
+    val seq: Seq[JsResult[JsValue]] = array.map(a => backReferenceResolver(serializer).reads(a)).toSeq
+    seq.find(res => res.isError)
+      .getOrElse(JsSuccess(JsArray(seq.map(res => res.get))))
+  }
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:98]: I think the indent was initially right?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala [hunk 2]
```diff
@@ -80,46 +86,67 @@ class ProcessingContext {
...
+      case _ => None
+    }).headOption
+
+    val transformedMap = resolutions.flatMap({
+      case Right((entry, value)) => Some((entry, value))
+      case _ => None
+    }).toMap
+
+    firstError.getOrElse(JsSuccess(JsObject(transformedMap)))
+  }
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:114]: indent here too

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala [hunk 3]
```diff
@@ -80,46 +86,67 @@ class ProcessingContext {
...
+
+  private def resolveBackReference(serializer: Serializer)(entry: (String, JsValue)): Either[JsError, (String, JsValue)] = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:116]: idem

**Comment by chibenwa** [comment]
Forced pushed as I did some significant changes:

 - Fixed some tiny mistakes preventing things to work...
 - JMAPApiRoutes does all the JSON translation and invocation recording work. Method developper just don't need to worry anymore about back-reference resolution (at all!)
 - I did a separated contracts and added a few tests regardings errors

I feel pretty happy about this work now, so if you agree with the overall strategy then I will write more tests and eventually undraft this!

**Comment by chibenwa** [comment]
Forced pushed to solve a simple conflict.

I added some tests for path resolution (works great!)

**Comment by chibenwa** [comment]
Deployment test failed

mvn tests :green_apple: 

test this please

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala [hunk 4]
```diff
@@ -19,18 +19,157 @@
...
+  private def readWildcard(jsValue: JsValue) = jsValue match {
+    case JsArray(arrayItems) =>
+      val evaluationResults: List[JsResult[JsValue]] = arrayItems.toList.map(evaluate)
+
+      evaluationResults.find(x => x.isInstanceOf[JsError])
+        .getOrElse(JsSuccess(expendArray(evaluationResults)))
+    case _ => JsError("Expecting an array")
+  }
+
+  private def expendArray(evaluationResults: List[JsResult[JsValue]]) = {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:71]: Yes I relied on IDE extraction, but did not added the types... Sorry.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala [hunk 5]
```diff
@@ -19,18 +19,157 @@
...
+    JsError(s"$resultOf references a ${invocation.methodName} method")
+  } else {
+    path.evaluate(invocation.arguments.value)
+  }
+}
+
+case class InvalidResultReferenceException(message: String) extends IllegalArgumentException
+
 class ProcessingContext {
  private val creationIds: mutable.Map[ClientId, ServerId] = mutable.Map()
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:92]: I don't understand.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala [hunk 6]
```diff
@@ -19,18 +19,157 @@
...
+ private val invocations: mutable.Map[MethodCallId, Invocation] = mutable.Map()
 
  def recordCreatedId(clientId: ClientId, serverId: ServerId): Unit = creationIds.put(clientId, serverId)
  private def retrieveServerId(clientId: ClientId): Option[ServerId] = creationIds.get(clientId)
 
+  def recordInvocation(invocation: Invocation): Unit = invocations.put(invocation.methodCallId, invocation)
+
+  def resolveBackReferences(serializer: Serializer, invocation: Invocation): Either[InvalidResultReferenceException, Invocation] =
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:100]: We only have one Serializer class. Do you mean renaming the "serializer" parameter to "deserializer"?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala [hunk 7]
```diff
@@ -114,13 +114,14 @@ class JMAPApiRoutes (val authenticator: Authenticator,
...
     val unsupportedCapabilities = requestObject.using.toSet -- DefaultCapabilities.SUPPORTED.ids
+    val capabilities: Set[CapabilityIdentifier] = requestObject.using.toSet
 
     if (unsupportedCapabilities.nonEmpty) {
       SMono.raiseError(UnsupportedCapabilitiesException(unsupportedCapabilities))
     } else {
       requestObject
         .methodCalls
-        .map(invocation => this.processMethodWithMatchName(requestObject.using.toSet, invocation, mailboxSession, processingContext))
+        .map(invocation => process(capabilities, mailboxSession, processingContext, invocation))
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala:124]: Contributions welcomed ;-)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/BackReferenceContract.scala [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+           |       "properties": ["id"]
+           |     },
+           |     "c1"],[
+           |     "Mailbox/get",
+           |     {
+           |       "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+           |       "properties": ["id"],
+           |       "#ids": {
+           |         "resultOf":"c42",
+           |         "name":"Mailbox/set",
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/BackReferenceContract.scala:427]:
```suggestion
           |         "name":"Mailbox/get",
```

**Review by chibenwa — COMMENTED** [review]
Nice thanks, I was trying to write something similar...

**Comment by chibenwa** [comment]
Ready for review <3

**Comment by chibenwa** [comment]
test this please (deployment test)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/BackReferenceContract.scala [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+         |                    },
+         |                    {
+         |                        "id": "3"
+         |                    },
+         |                    {
+         |                        "id": "4"
+         |                    },
+         |                    {
+         |                        "id": "6"
+         |                    }
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/BackReferenceContract.scala:109]: Wouldn't it be less heavy and acceptable to write each items in the list like this, as we only have id anyway: `{"id": "6"},` ?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala [hunk 10]
```diff
@@ -129,12 +130,25 @@ class JMAPApiRoutes (val authenticator: Authenticator,
...
                 serializer.serialize(ResponseObject(ResponseObject.SESSION_STATE, invocations)).toString),
-              StandardCharsets.UTF_8
-            ).`then`())
+              StandardCharsets.UTF_8)
+            .`then`())
         )
     }
   }
 
+  private def process(capabilities: Set[CapabilityIdentifier], mailboxSession: MailboxSession, processingContext: ProcessingContext, invocation: Invocation) = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala]: No return type defined?

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala]: This comment is outdated

**Review by Arsnael — APPROVED** [review]

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala [hunk 11]
```diff
@@ -19,18 +19,155 @@
...
+    }
+
+  private def backReferenceResolver(serializer: Serializer): Reads[JsValue] = {
+    case JsArray(value) => resolveBackReferences(serializer, value)
+    case JsObject(underlying) => resolveBackReference(serializer, underlying)
+    case others: JsValue => JsSuccess(others)
+  }
+
+  private def resolveBackReferences(serializer: Serializer, array: collection.IndexedSeq[JsValue]): JsResult[JsValue] = {
+    val seq: Seq[JsResult[JsValue]] = array.map(a => backReferenceResolver(serializer).reads(a)).toSeq
```
Comment (nvduc91) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/ProcessingContext.scala:114]: `a` could is had better name?

**Review by nvduc91 — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
`RabbitMQEventBusTest$ConcurrentTest.concurrentDispatchKeyShouldDeliverAllEventsToListenersWithSingleEventBus » RetryExhausted`

unrelated

test this please

**Review by remk — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

deployment test - can be considered :green_apple:

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**